src/bun.js/net.exports.js

```diff
@@ -53,9 +53,257 @@ export function isIP(s) {
+      }
+    }
+  }
+
+  bytesRead = 0;
+  bytesWritten = 0;
+  #closed = false;
+  connecting = false;
+  localAddress = "127.0.0.1";
+  #readQueue = createFIFO();
+  remotePort;
+  #socket;
+  timeout = 0;
+  #writeCallback;
+  #writeChunk;
+
+  constructor(options) {
+    super({
+      allowHalfOpen: options?.allowHalfOpen || false,
+      readable: true,
+      writable: true,
+    });
+    options?.signal?.once("abort", () => this.destroy());
+    this.once("connect", () => this.emit("ready"));
+    // TODO support `options.fd`
+  }
+
+  address() {
+    return {
+      address: this.localAddress,
+      family: this.localFamily,
+      port: this.localPort,
+    };
+  }
+
+  get bufferSize() {
+    return this.writableLength;
+  }
+
+  connect(port, host, connectListener) {
+    // TODO support IPC sockets
+    if (typeof host == "function") {
+      connectListener = host;
+      host = undefined;
+    }
+    if (typeof port == "object") {
+      var {
+        port,
+        host,
+        // TODOs
+        localAddress,
+        localPort,
+        family,
+        hints,
+        lookup,
+        noDelay,
+        keepAlive,
+        keepAliveInitialDelay,
+      } = port;
+    }
+    this.connecting = true;
+    this.remotePort = port;
+    if (connectListener) this.on("connect", connectListener);
+    bunConnect({
+      data: this,
+      hostname: host || "localhost",
+      port: port,
+      socket: Socket.#Handlers,
+    });
+    return this;
+  }
+
+  _destroy(err, callback) {
+    this.#socket?.end();
+    callback(err);
+  }
+
+  _final(callback) {
+    this.#socket.end();
+    callback();
+  }
+
+  get localAddress() {
+    return "127.0.0.1";
+  }
+
+  get localFamily() {
+    return "IPv4";
+  }
+
+  get localPort() {
+    return this.#socket?.localPort;
+  }
+
+  get pending() {
+    return this.connecting;
+  }
+
+  _read(size) {
+    const queue = this.#readQueue;
+    let chunk;
+    while (chunk = queue.peek()) {
+      if (!this.push(chunk)) break;
+      queue.shift();
+    }
+  }
+
+  get readyState() {
+    if (this.connecting) return "opening";
+    if (this.readable) {
+      return this.writable ? "open" : "readOnly";
+    } else {
+      return this.writable ? "writeOnly" : "closed";
+    }
+  }
+
+  ref() {
+    this.#socket?.ref();
+  }
+
+  get remoteAddress() {
+    return this.#socket.remoteAddress;
+  }
+
+  get remoteFamily() {
+    return "IPv4";
+  }
+
+  resetAndDestroy() {
+    this.#socket?.end();
+  }
+
+  setKeepAlive(enable = false, initialDelay = 0) {
+    // TODO
+  }
+
+  setNoDelay(noDelay = true) {
+    // TODO
+  }
+
+  setTimeout(timeout, callback) {
+    this.#socket?.timeout(timeout);
+    this.timeout = timeout;
+    if (callback) this.once("timeout", callback);
+    return this;
+  }
+
+  unref() {
+    this.#socket?.unref();
+  }
+
+  _write(chunk, encoding, callback) {
+    if (typeof chunk == "string" && encoding !== "utf8") chunk = Buffer.from(chunk, encoding);
```

Comment:
If we split this into two functions it _might_ be faster:

1) if (typeof chunk === "string") this.#writeString(chunk, callback)
2) this.#writeBuffer(chunk, callback)

We could also add a DOMJIT binding for writeText / writeBytes, as we have done for ServerWebSocket and then we can have lower function call overhead in these functions. It would only shave off around 20ns though

Reply:
Just to clarify, the above code is there to handle things like `socket.setEncoding("hex").write("deadbeef")`, i.e. for correctness rather than performance